### ISSUE TITLE
npm install error: cairo not found

### ISSUE THREAD

**Issue by mojoaxel** [almende/vis#3515]
Since #3262 i get the following error trying `npm install` on linux:

```sh
> canvas@1.6.7 install /home/delphin/workspaces/workspace_vis/vis/node_modules/canvas
> node-gyp rebuild

gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
gyp WARN download NVM_NODEJS_ORG_MIRROR is deprecated and will be removed in node-gyp v4, please use NODEJS_ORG_MIRROR
Package cairo was not found in the pkg-config search path.
Perhaps you should add the directory containing `cairo.pc'
to the PKG_CONFIG_PATH environment variable
No package 'cairo' found
gyp: Call to './util/has_lib.sh freetype' returned exit status 0 while in binding.gyp. while trying to load binding.gyp
...
```

**Comment by wimrijnders** [comment]
@mojoaxel I saw the assignment. How can I solve this if it works on my linux? I would think the best person to fix it is you, since it happens on yourr.

**Comment by wimrijnders** [comment]
How about disabling the specific unit tests of the canvas is not present under `node.js`? This would at least prevent problems with running them locally.

**Comment by mojoaxel** [comment]
> perhaps they did a manual install of cairo?

Of course this fixes the problem. But still this is an external dependency. How do you plan to communicated this to the user?

As far as I am concerned we should absolut not use `node-canvas`as long it has this external dependency. As long as there is no `node-cairo` I think we should look around for other test-solutions.

**Comment by wimrijnders** [comment]
I disagree. It's supremely useful for the unit tests and I consider those very important.

But I am willing to consider `canvas` alternatives for `node.js`, do you know any?

**Comment by wimrijnders** [comment]
>  But still this is an external dependency. 

There are setups which can handle the dependency automatically. How about checking out how that works? See also link above about `pkdconfig`.

**Comment by mojoaxel** [comment]
> How about disabling the specific unit tests of the canvas is not present under node.js? This would at least prevent problems with running them locally.

This could be a workaround, but developers still must be informed that Cairo is an requirement to develop test. To be honest i'm not happy with this dependency but to be honest i also do not have the time at the moment to develop a better test solution. 😿 

I propose: Let's treat this issue as a blocker (no new releases until we found a solution) and think about how to fix this. My preferred solution would be to use a [ui-testing framework](https://github.com/mojoaxel/awesome-regression-testing) that did not have any non-npm dependencies, but as I said time is an issue...

**Comment by wimrijnders** [comment]
> This could be a workaround, but developers still must be informed that Cairo is an requirement to develop test.

This can be combined. Disable the tests by default, in the README explain that in order to run them, you will need to install cairo. That way, you have a non-blocking setup that can be extended.

> ... i also do not have the time at the moment to develop a better test solution. 😿

That's because there isn't any. I was hoping you would take the hint.

> Let's treat this issue as a blocker

No. But see my solution at the top of this comment.

**Comment by wimrijnders** [comment]
**everyone:* I'd like to know if you installed `cairo` by hand to get the unit tests to work. Is this issue also an issue for you?

**Comment by mojoaxel** [comment]
>  See also link above about pkdconfig.

As I sayed: Installing Cairo ([as descripe here](https://github.com/Automattic/node-canvas/wiki/Installation---Ubuntu-and-other-Debian-based-systems)) would solve this issue for me, of course. But I'm very sure this is a big issue for anybody who wants to develop vis.js and I really don't think that this kind of dependencies are a good idea!

**Comment by wimrijnders** [comment]
Your point is noted; however, consider all the things you need to do to develop for `vis.js` in the first place. In the big picture, it's not such a big deal.

Is the solution, interim or not, of disabling cairo-based tests accetable? And that they *will* run if cairo is installed?

**Comment by mojoaxel** [comment]
> Is the solution, interim or not, of disabling cairo-based tests accetable? And that they will run if cairo is installed?

Yes of course! I'm absolutely ok with me treating this cairo-based tests as a interims solution untill we have a better testing framework. But than the cairo dependency, and how to install it, must be really well documented.

**Comment by wimrijnders** [comment]
Documenting this is something which I do not contest; I hope that is clear.

What I do react against is the implication that `cairo` should not be used at all and that production should be blocked until this is somehow resolved. IMHO, way over the top.

*And* the fact that you seem to be the only having a problem with installing cairo and feel that this is sufficient to make all the above statements. But I am waiting to be corrected by others on this point.

**Comment by yotamberk** [comment]
I haven't had any problems with Cairo.
I'm with @wimrijnders about not blocking it. We can explain that for now, in-order of running tests, Cairo is required. I really don't think it's a show stopper that requires stopping any releases.

**Comment by mojoaxel** [comment]
OK, that please add documentation that Cairo is now a requirement for vis.js development.
I'll see that I find a better solution for the future...

**Comment by wimrijnders** [comment]
Thanks for support. I've cooled down a bit now. While I don't think it's a show-stopper either, I appreciate that @mojoaxel's core points are valid. My interpretation:

- The unit tests should run out of the box with a default install
- Developers should be able to find out what to do to enable full unit tests

I believe that this is the least intrusive. My consideration are:
-  Users of `vis.js` do not have to deal with this at all (duh).
-  Any developer not working with `Network` does not have to do anything special.
-  Any developer working with `Network` may or may not have to enable them. It just depends on what they're working on
- Travis has no problems at all with running cairo. This is the most important for me

Still I would like to know from more people if they had to manually install `cairo`.

**Comment by justinharrell** [comment]
On Windows I had to manually install GTK into c:\GTK before Cairo would npm install which to took me some time to track down and figure out, I also got lucky because I already have Visual Studio installed, without that I don't believe it will build.

I eventually tracked down the docs for windows here: https://github.com/Automattic/node-canvas/wiki/Installation---Windows

This was not a great experience since there was no mention of it in the docs and a little frustrating to have a native module dependency blocking me from working on vis that was only used in the test framework not even needed for the library itself.

I always dread having to deal with a native node module on windows, getting it to build is hardly ever painless.

**Comment by wimrijnders** [comment]
The keyword here is 'frustrating'. OK, noted. In fact, having just one more person dealing with this is enough motivation for me to take preventive measures.

Anyone else had a bad time with cairo?

**Comment by wimrijnders** [comment]
OK figured out how to skip tests if `canvas` not installed. In the main describe-block of the `Network` unit tests:

```js
describe('Network', function () {
  before(function() {
...
    var ctx = window.document.createElement('canvas').getContext('2d');
    this.canvasPresent = (typeof ctx !== 'undefined' && ctx !== null);

    if (!this.canvasPresent) this.skip();
...
  });
...
```

And in every contained describe block:
```js
describe('Network', function () {
  before(function() {
    if (!this.canvasPresent) this.skip();
  });
...
```

The interesting thing here is that the unit tests are still displayed, but with a different color.
This is actually not a bad thing; it's an indication that something out of the ordinary is going on.

Then, with some strategically placed documentation, a developer can figure out for his/herself WTF is happening. And then know if this is something to worry about or not.

----
Next up: how to not installl `canvas` by default, except when running on `Travis`. Suggestions are welcome.

**Comment by justinharrell** [comment]
If you are not worried about testing the canvas output, something like this would be much lightier with no native dependencies: https://github.com/Cristy94/canvas-mock

**Comment by wimrijnders** [comment]
Thanks for link, will seriously consider that. However, some minimal behavior should be present.
OTOH, it's really basic. It's not overkill to copy over the idea and suit it for own purposes.

**Comment by justinharrell** [comment]
https://github.com/gliffy/canvas2svg might also be an option that actually does produce an artifact that can be viewed. It does mention running in node and may not need node-canvas with newer versions of jsdom and it has not native dependencies.

**Comment by wimrijnders** [comment]
I take my hat off to this plug, which as it happens is entirely appropriate. You see a way to make it work?

**Comment by wimrijnders** [comment]
Come on guys! One more person with a bad experience on installing cairo, and I'll put back the `Priority` label myself. Bring it on!

**Comment by justinharrell** [comment]
I will not lie it would make me happy to see vis.js be dependent on canvas2svg, it just so happens one side effect of doing svg is you don't need a native drawing library to produce it, its just text :).

I don't fully understand why they mention node-canvas being needed on older jsdoms and recommend  otherwise, that would be the main issue to figure out, other I already know canvas2svg works pretty well with vis.js.

**Comment by wimrijnders** [comment]
I am fully prepared to go with your project, if only because you've amply proven to me that you're a capable developer. Your PR's thus far inspire plenty confidence.

From another viewpoint, we need to make it work anyway for the upcoming SVG export, so it doesn't feel like additional work to me. I suppose clearing out the drawing routines grafted into canvas (`shapes.js`) applies here as well?

Hm. I'm jumping the gun here as usual. I propose first getting this, err, 'mess' cleared out. Then get the svg export on the roll, then see if it can be used for transplanting `canvas`. I'm warming up to the canvas-dummy idea now. I've actually been using something of the sort for 2d-context in the `Label` unit tests.

**Comment by clayt85** [comment]
> Come on guys! One more person with a bad experience on installing cairo, and I'll put back the `Priority` label myself. Bring it on!

I'm late to the game (one year later) but yes, this was a problem for me. Happened with both CentOS 6 and Fedora 27.
